watertap/unit_models/nanofiltration_0D.py

```diff
@@ -0,0 +1,732 @@
+                b.properties_permeate[t].get_material_flow_terms(p, j)
+                for p in pset
+                if (p, j) in pcset
+            )
+
+        @self.Constraint(self.flowsheet().time, self.properties_in.phase_component_set)
+        def separation_constraint(b, t, p, j):
+            comp = prop_params.get_component(j)
+
+            if comp.is_solvent():
+                # Permeate flows equal to recovery * inlet flow
+                return b.solvent_recovery[t] * b.properties_in[
+                    t
+                ].get_material_flow_terms(p, j) == b.properties_permeate[
+                    t
+                ].get_material_flow_terms(
+                    p, j
+                )
+            elif j == self.config.electroneutrality_ion:
+                # Rejection of electroneutrality ion will be solved using electroneutrality constraint
+                return Constraint.Skip
+
+            # else: Rejection = 1 - C_permeate/C_feed
+            return (
+                (1 - b.rejection_comp[t, j])
+                * b.properties_in[t].conc_mol_phase_comp[p, j]
+            ) == b.properties_permeate[t].conc_mol_phase_comp[p, j]
+
+        if self.config.electroneutrality_ion is not None:
+            # Add electroneutrality constraint for permeate stream
+            @self.Constraint(self.flowsheet().time)
+            def permeate_electronegativity(b, t):
+                perm_state = b.properties_permeate[t]
+                return 0 == sum(
+                    perm_state.params.get_component(j).config.charge
+                    * perm_state.flow_mol_phase_comp[p, j]
+                    for p, j in perm_state.phase_component_set
+                    if j in perm_state.params.ion_set
+                )
+
+        # Retentate pressure balance
+        if self.config.momentum_balance_type == MomentumBalanceType.pressureTotal:
+
+            @self.Constraint(self.flowsheet().time, doc="Retentate pressure balance")
+            def retentate_pressure_balance(b, t):
+                expr = b.properties_retentate[t].pressure
+                if b.config.has_pressure_change:
+                    expr += -b.deltaP[t]
+                return b.properties_in[t].pressure == expr
+
+        # Temperature equalities
+        if self.config.energy_balance_type == EnergyBalanceType.isothermal:
+
+            @self.Constraint(
+                self.flowsheet().time, doc="Retentate temperature equality"
+            )
+            def retentate_temperature_equality(b, t):
+                return (
+                    b.properties_in[t].temperature
+                    == b.properties_retentate[t].temperature
+                )
+
+            @self.Constraint(self.flowsheet().time, doc="Permeate temperature equality")
+            def permeate_temperature_equality(b, t):
+                return (
+                    b.properties_in[t].temperature
+                    == b.properties_permeate[t].temperature
+                )
```

Comment:
Any reason not to just apply the modified control volume? I was assuming that it was written in a way that one wouldn't need to write out the equality constraints (but been a while since last reviewed new control volume).

Reply:
Mainly because this model does not use any control volumes (but does require the `EnergyBalanceType.isothermal` option in the Enum). In short, as this is basically a specialised separator a control volume does not add much as we would need to write at least half the constraints manually anyway.